File: nanostream.zig
const std = @import("std");

// nanostream: unaligned bit writer and reader utility

pub const BitWriter = struct {
    buf: std.ArrayListUnmanaged(u8),
    alloc: std.mem.Allocator,
    current: u64 = 0,
    /// number of bits written in `current`: 0–63
    bitIndex: u8 = 0,

    /// recommended `capacity` is 4.
    pub fn init(alloc: std.mem.Allocator, capacity: usize) error{OutOfMemory}!BitWriter {
        return .{
            .buf = try .initCapacity(alloc, capacity),
            .alloc = alloc,
        };
    }

    /// Writes bits and moves forward.
    ///
    /// - `stuff`: Expects LTR structure placed at the right like `0b00010100`
    /// - `bits`: Number of bits to write from `stuff` (RTL). Maximum `64`. I can't use comptime here, so I'm just trusting you DO NOT MAKE IT >64.
    ///
    /// Ex: `stuff`=`0b0000000000000000000000000000000000000000000000000000000011110100` with bits = `5` adds `10100` to the stream.
    fn writeBits(self: *BitWriter, stuff: u64, bits: u8) void {
        var stufff = stuff; // make it mut

        // clear the top bytes we arent writing
        if (bits < 64) {
            stufff &= (@as(u64, 1) << @intCast(bits)) - @as(u64, 1);
        }

        if (self.bitIndex + bits >= 64) {
            // write in 2 parts
            const bits_1: u6 = @truncate(64 - self.bitIndex);
            const bits_2: u6 = @truncate(bits - (64 - self.bitIndex));

            const stuff_1: u64 = stufff >> bits_2;
            self.current <<= bits_1;
            self.current |= stuff_1;
            // == ByteList.Add_u64 ==
            self.buf.ensureUnusedCapacity(self.alloc, 8) catch unreachable; // this will crash on OOM — feel free to do something different, but who wants to "try w.write" every time
            const dst: *[8]u8 = self.buf.unusedCapacitySlice()[0..8];
            std.mem.writeInt(u64, dst, self.current, .big);
            self.buf.items.len += 8;
            // ======================

            self.current = 0;
            if (bits_2 > 0) {
                const stuff_2 = stufff & ((@as(u64, 1) << bits_2) - @as(u64, 1));
                self.current <<= bits_2;
                self.current |= stuff_2;
            }
        } else {
            self.current <<= @intCast(bits);
            self.current |= stufff;
        }

        self.bitIndex += bits;
        self.bitIndex %= 64;
    }
    /// - bits: Finishes up the current byte. If this completes the u64 `current`, `WriteBits` automatically flushes the entire u64 it into the array.
    /// - bytes: In case we didn't finish the u64 `current`, flush it byte-by-byte anyway (not optimized, we don't write a whole u64).
    fn flushBitsAndBytes(self: *BitWriter) error{OutOfMemory}!void {
        if (self.bitIndex > 0) {
            // finish up current byte exactly
            if (self.bitIndex % 8 != 0)
                self.writeBits(0, 8 - (self.bitIndex % 8));

            // force push single bytes (if we added a u64 already: bitIndex == 0 -> skip)
            // bitIndex must be a multiple of 8 after WriteBits
            while (self.bitIndex > 0) {
                try self.buf.append(self.alloc, @as(u8, @truncate(self.current >> @intCast(self.bitIndex - 8))));
                self.bitIndex -= 8;
            }
        }
    }

    /// Finalization function.
    /// Force write the current byte, which isn't normally written until it reaches 8 bits. Call after writing all bits.
    ///
    /// Warning: you now own this slice I give you, free it later with the allocator you gave me! (unless you like leaking)
    pub fn conclude(self: *BitWriter) error{OutOfMemory}![]u8 {
        try self.flushBitsAndBytes();
        return self.buf.toOwnedSlice(self.alloc);
    }

    pub fn write(self: *BitWriter, n: anytype) void {
        switch (@typeInfo(@TypeOf(n))) {
            .int => |info| {
                if (info.signedness == .signed) {
                    writeBits(self, @bitCast(@as(i64, n)), info.bits);
                } else {
                    writeBits(self, @as(u64, n), info.bits);
                }
            },
            .float => |info| {
                switch (info.bits) {
                    16 => writeBits(self, @as(u16, @bitCast(@as(f16, n))), 16),
                    32 => writeBits(self, @as(u32, @bitCast(@as(f32, n))), 32),
                    64 => writeBits(self, @as(u64, @bitCast(@as(f64, n))), 64),

                    else => @compileError("I CANT WRITE TYPE '" ++ @typeName(@TypeOf(n)) ++ "'"),
                }
            },
            .bool => {
                writeBits(self, @intFromBool(n), 1);
            },
            .@"enum" => |info| {
                const backing_val = @intFromEnum(n);
                const backing_info = @typeInfo(info.tag_type).int;
                if (backing_info.signedness == .signed) {
                    self.writeBits(@bitCast(@as(i64, backing_val)), backing_info.bits);
                } else {
                    self.writeBits(@as(u64, backing_val), backing_info.bits);
                }
            },
            .@"struct" => |info| {
                if (info.layout == .auto) {
                    @compileError("struct '" ++ @typeName(@TypeOf(n)) ++ "' must be `extern` or `packed` for a stable wire layout");
                }
                self.writeByteArray(std.mem.asBytes(&n));
            },

            // ... (all iX, uX, fX, plus bool. the custom types like i27 works as well probably)
            else => {
                @compileError("I CANT WRITE TYPE '" ++ @typeName(@TypeOf(n)) ++ "'");
            },
        }
    }

    pub fn writeString(self: *BitWriter, str: []const u8) void {
        self.write(@as(u16, @intCast(str.len)));
        self.writeByteArray(str);
    }

    fn writeByteArray(self: *BitWriter, n: []const u8) void {
        var ptr = n;
        while (ptr.len >= 8) : (ptr = ptr[8..]) {
            self.writeBits(std.mem.readInt(u64, ptr[0..8], .big), 64);
        }
        while (ptr.len > 0) : (ptr = ptr[1..]) {
            self.writeBits(ptr[0], 8);
        }
    }
};

pub const BitReader = struct {
    buf: []const u8,
    byteIndex: u32 = 0,
    cache: u64 = 0,
    bitIndex: u8 = 0,

    pub fn init(buf: []const u8) BitReader {
        var r: BitReader = .{
            .buf = buf,
        };
        r.obtainCache();

        return r;
    }

    fn obtainCache(self: *BitReader) void {
        if (self.byteIndex + 8 <= self.buf.len) {
            self.cache = std.mem.readInt(u64, self.buf[self.byteIndex..][0..8], .big);
        } else {
            self.cache = 0;
            var idx: u32 = self.byteIndex;
            while (idx < self.buf.len) {
                self.cache <<= 8;
                self.cache |= self.buf[idx];
                idx += 1;
            }

            self.cache <<= @as(u6, @truncate(64 - 8 * (self.buf.len - self.byteIndex)));
        }
    }

    /// Reads bits and moves forward.
    ///
    /// - `bits`: Maximum `64`.
    ///
    /// Bits are placed in an output of the exact specified bit-width.
    /// Ex. ...`0000000`(You are here)`1100101`... with `bits`=`5` returns a `u5`: `0b11001`.
    fn readBits(self: *BitReader, comptime bits: u8) @Int(.unsigned, bits) {
        if (bits > 64) @compileError("bits can't be >64 buddy");
        if (bits == 0) return 0;

        var output: u64 = 0;

        if (self.bitIndex + bits >= 64) {
            const bits_1: u8 = @intCast(64 - self.bitIndex);
            const bits_2: u8 = @intCast(bits - bits_1);

            const reading_1 = self.cache >> @intCast(64 - bits_1);

            self.byteIndex += 8;
            self.obtainCache();

            if (bits_2 > 0) {
                const reading_2: u64 = self.cache >> @intCast(64 - bits_2);
                self.cache <<= @intCast(bits_2);
                output = (reading_1 << @intCast(bits_2)) | reading_2;
            } else {
                output = reading_1;
            }
        } else {
            output = self.cache >> @intCast(64 - bits);
            self.cache <<= @truncate(bits); // wont compile without @truncate. safe since we know bits<64 in this path. compiler dumb sometimes lol (it should know bits<64 too)
        }

        self.bitIndex +%= bits;
        self.bitIndex %= 64;
        return @truncate(output);
    }

    pub fn read(self: *BitReader, comptime T: type) T {
        switch (@typeInfo(T)) {
            .int => |info| {
                return @bitCast(self.readBits(info.bits));
            },
            .float => |info| {
                if (!(info.bits == 16 or info.bits == 32 or info.bits == 64)) @compileError("I CANT READ TYPE '" ++ @typeName(T) ++ "'");
                return @bitCast(self.readBits(info.bits));
            },
            .bool => {
                return self.readBits(1) == 1;
            },

            // holy im cooking
            .@"enum" => |info| {
                const backing_info = @typeInfo(info.tag_type).int;
                const xchg: @Int(backing_info.signedness, backing_info.bits) = @bitCast(self.readBits(backing_info.bits));
                return @enumFromInt(xchg);
            },

            .@"struct" => |info| {
                if (info.layout == .auto) {
                    @compileError("struct '" ++ @typeName(T) ++ "' must be `extern` or `packed` for a stable wire layout");
                }

                var value: T = undefined;
                self.readByteArray(std.mem.asBytes(&value));
                return value; // this is pass-by-value, no alloc needed
            },

            else => {
                @compileError("I CANT READ TYPE '" ++ @typeName(T) ++ "'");
            },
        }
    }

    /// Warning: you own this string now, free it later with the allocator you gave me! (unless you like leaking)
    ///
    /// string may be null, for interop with other languages
    pub fn readString(self: *BitReader, alloc: std.mem.Allocator) error{OutOfMemory}!?[]u8 {
        const len = self.read(u16);
        if (len == std.math.maxInt(u16)) {
            return null;
        }
        if (len == 0) {
            return &[0]u8{};
        }
        const slice: []u8 = try alloc.alloc(u8, len);
        self.readByteArray(slice);
        return slice;
    }

    fn readByteArray(self: *BitReader, slice: []u8) void {
        var ptr = slice; // make it mut

        // slice is a fat pointer, so it contains both ptr and len from the C# version
        while (ptr.len >= 8) : (ptr = ptr[8..]) {
            std.mem.writeInt(u64, ptr[0..8], self.readBits(64), .big);
        }
        while (ptr.len > 0) : (ptr = ptr[1..]) {
            ptr[0] = self.readBits(8);
        }
    }
};

test "symmetry" {
    const Thing = packed struct {
        f: f32,
        b: bool,
        u: u16,
    };
    const Thing2 = extern struct {
        f: f32 align(1),
        i: i64 align(256),
        u: u32 align(1),
    };

    var gpa = std.heap.DebugAllocator(.{}){};
    const DBG = gpa.allocator();
    defer _ = gpa.deinit();

    // write
    var w = try BitWriter.init(DBG, 1024);

    const a: u8 = 43;
    const b: bool = false;
    const c: i32 = -33337;
    const d: f64 = 71.4;
    const e: u63 = 4328948392435433;
    const f = "wazzup beijing";
    const g = Thing{ .f = -10, .b = false, .u = 5000 };
    const h = Thing2{ .f = -3249.51, .i = 88942121, .u = 400000000 };

    w.write(a);
    w.write(b);
    w.write(c);
    w.write(d);
    w.write(e);
    w.writeString(f);
    w.write(g);
    w.write(h);

    const arr = try w.conclude();
    defer DBG.free(arr);

    // read
    var r = BitReader.init(arr);
    try std.testing.expectEqual(r.read(u8), a);
    try std.testing.expectEqual(r.read(bool), b);
    try std.testing.expectEqual(r.read(i32), c);
    try std.testing.expectEqual(r.read(f64), d);
    try std.testing.expectEqual(r.read(u63), e);
    const fr = (try r.readString(DBG)).?;
    defer DBG.free(fr);
    try std.testing.expectEqualStrings(f, fr);
    try std.testing.expectEqual(r.read(Thing), g);
    try std.testing.expectEqual(r.read(Thing2), h);
}

File: Nanostream.cs

using System;
using System.Runtime.CompilerServices;
using static System.Runtime.CompilerServices.MethodImplOptions;
using System.Buffers.Binary;

// nanostream: unaligned bit writer and reader utility
namespace Nanostream
{
    /// <summary>
    /// Continuous bit writer. (i don't pad 0s, screw alignment)<br/>
    /// </summary>
    public class BitWriter
    {
        public BitWriter(int capacity = 4) { buf = new(capacity); current = 0; bitIndex = 0; }
        // public BitWriter(ByteList bytes) { buf = bytes; current = 0; bitIndex = 0; }
        // public void WriteBytesAligned(ReadOnlySpan<byte> bytes)
        // {
        //     if (bitIndex % 8 != 0) throw new InvalidOperationException("BitWriter must be aligned to write a span of bytes!");
        //     TRYFinishCurrentByteANDForcePush();
        //     buf.AddRange(bytes);
        // }
        ByteList buf;

        ulong current;
        ///<summary>number of bits written in <c>current</c>: 0–63</summary>
        byte bitIndex;

        /// <summary>
        /// Writes bits and moves forward.
        /// </summary>
        /// <param name="stuff">Expects LTR structure placed at the right like <c>0b00010100</c></param>
        /// <param name="bits">Number of bits to write from <c>stuff</c> (RTL).<br/>Maximum <c>64</c>.<br/>Ex: <c>stuff</c>=<c>0b0000000000000000000000000000000000000000000000000000000011110100</c> with bits = <c>5</c> adds <c>10100</c> to the stream.</param>
        // should test [MethodImpl(AggressiveInlining)], might drastically increase build size since so many Write() calls exist everywhere.
        void WriteBits(ulong stuff, byte bits = 8)
        {
            // clear the top bytes we arent writing
            if (bits < 64)
                stuff &= (1ul << bits) - 1ul;

            if (bitIndex + bits >= 64)
            {
                // write in 2 parts
                byte bits_1 = (byte)(64 - bitIndex);
                byte bits_2 = (byte)(bits - bits_1);

                ulong stuff_1 = stuff >> bits_2;
                current <<= bits_1;
                current |= stuff_1;
                buf.Add_u64(current);

                current = 0;
                if (bits_2 > 0)
                {
                    ulong stuff_2 = stuff & ((1ul << bits_2) - 1ul);
                    current <<= bits_2;
                    current |= stuff_2;
                }
            }
            else
            {
                // fits into current ulong
                current <<= bits;
                current |= stuff;
            }

            bitIndex += bits;
            bitIndex %= 64;
        }
        /// <summary>
        /// - bits: Finishes up the current byte. If this completes the u64 <c>current</c>, <c>WriteBits</c> automatically flushes the entire u64 it into the array.<br/>
        /// - bytes: In case we didn't finish the u64 <c>current</c>, flush it byte-by-byte anyway (not optimized, we don't write a whole u64).
        /// </summary>
        void FlushBitsAndBytes()
        {
            if (bitIndex > 0)
            {
                // finish up current byte exactly
                if (bitIndex % 8 != 0)
                    WriteBits(0, (byte)(8 - (bitIndex % 8)));

                // force push single bytes (if we added a u64 already: bitIndex == 0 -> skip)
                // bitIndex must be a multiple of 8 after WriteBits
                while (bitIndex > 0)
                {
                    buf.Add((byte)(current >> (bitIndex - 8)));
                    bitIndex -= 8;
                }
            }
        }
        /// <summary>
        /// Finalization function.
        /// Force write the current byte, which isn't normally written until it reaches 8 bits. Call after writing all bits.
        /// </summary>
        public byte[] Conclude()
        {
            FlushBitsAndBytes();
            return buf.ToArray();
        }
        [MethodImpl(AggressiveInlining)] public void Write(sbyte i) => Write((byte)i); // cast preserves the sign bit
        [MethodImpl(AggressiveInlining)] public void Write(short i) => Write((ushort)i); // cast preserves the sign bit
        [MethodImpl(AggressiveInlining)] public void Write(int i) => Write((uint)i);
        [MethodImpl(AggressiveInlining)] public void Write(long i) => Write((ulong)i);

        [MethodImpl(AggressiveInlining)]
        public void Write(byte i)
        {
            WriteBits(i, 8);
        }
        [MethodImpl(AggressiveInlining)]
        public void Write(ushort i)
        {
            WriteBits(i, 16);
        }
        [MethodImpl(AggressiveInlining)]
        public void Write(uint i)
        {
            WriteBits(i, 32);
        }
        [MethodImpl(AggressiveInlining)]
        public void Write(ulong i)
        {
            WriteBits(i, 64);
        }
        [MethodImpl(AggressiveInlining)] public void Write(bool i) => WriteBits(i ? (byte)1 : (byte)0, 1);
        [MethodImpl(AggressiveInlining)] public void Write(float i) => Write(System.BitConverter.SingleToInt32Bits(i));
        [MethodImpl(AggressiveInlining)] public void Write(double i) => Write(System.BitConverter.DoubleToInt64Bits(i));

        // [MethodImpl(AggressiveInlining)] public void Write(Vector2 i) { Write(i.x); Write(i.y); }

        // public void Write(Fix64 i) { Write(i.RawValue); }

        /// <summary>my Unsafe Code cannot be this Peak</summary>
        [MethodImpl(AggressiveInlining)]
        public unsafe void Write<TE>(TE value) where TE : unmanaged, Enum
        {
            int size = Unsafe.SizeOf<TE>();
            void* ptr = &value;
            ulong val = size switch
            {
                1 => *(byte*)ptr,
                2 => *(ushort*)ptr,
                4 => *(uint*)ptr,
                8 => *(ulong*)ptr,
            };
            WriteBits(val, (byte)(size * 8));
        }
        
        /// <summary>
        /// Write any unmanaged struct (size never changes; no List, only fixed arrays).<br/><br/>
        /// <b>Warning</b>: with should work on-the-wire and be compatible with Zig's read(TS), but it's untested.<br/>
        /// </summary>
        [MethodImpl(AggressiveInlining)]
        public unsafe void WriteStruct<TS>(TS value) where TS : unmanaged
        {
            byte* ptr = (byte*)&value;
            WriteByteArrayUnsafe(ptr, (ushort)sizeof(TS));
        }

        /// <summary>
        /// I'm not gonna recommend this for performance (if string is long).<br/>
        /// A big string is already a byte[], why use an unaligned bit writer for that? Just send that first.<br/><br/>
        /// <b>Warning</b>: If <c>s</c> is 65535 bytes or longer, I will write a null string instead.
        /// </summary>
        [MethodImpl(AggressiveInlining)]
        public unsafe void WriteString(string s)
        {
            if (s == null)
            {
                Write(ushort.MaxValue);
                return;
            }
            byte[] bytes = System.Text.Encoding.UTF8.GetBytes(s);
            if(bytes.Length >= ushort.MaxValue)
            {
                Write(ushort.MaxValue);
                return;
            }

            fixed (byte* ptr = bytes)
            {
                byte* p = ptr;
                Write((ushort)bytes.Length);
                WriteByteArrayUnsafe(p, bytes.Length);
            }
        }

        /// <summary>
        /// Initially not public API, don't use this unless you know what you are doing!<br/>
        /// Requires a managed reference to be "pinned" to use pointers. If you have a <c>byte[]</c>, use a <c>fixed(byte* ptr = bytes)</c> statement.
        /// </summary>
        [MethodImpl(AggressiveInlining)]
        public unsafe void WriteByteArrayUnsafe(byte* ptr, int len)
        {
            int remaining = len;
            while (remaining >= 8)
            {
                WriteBits(BinaryPrimitives.ReadUInt64BigEndian(new ReadOnlySpan<byte>(ptr, 8)), 64);
                ptr += 8;
                remaining -= 8;
            }
            while (remaining > 0)
            {
                WriteBits(*ptr, 8);
                ptr += 1;
                remaining -= 1;
            }
        }
    }

    /// <summary>
    /// Continuous bit reader. (warning: ensure your schemas match, dont read extra.)
    /// </summary>
    public class BitReader
    {
        // use this constructor to make a copy of a BitReader at a certain position in case you need to re-read stuff later (like after gaining info from later on the stream)
        public BitReader(BitReader snapshot)
        {
            this.buf = snapshot.buf; this.byteIndex = snapshot.byteIndex; this.bitIndex = snapshot.bitIndex;
            this.cache = snapshot.cache;
        }
        public BitReader(byte[] buf)
        {
            this.buf = buf; byteIndex = 0; bitIndex = 0;
            ObtainCache();
        }
        private byte[] buf;

        uint byteIndex;
        ulong cache;
        ///<summary>number of bits read in <c>cache</c>: 0–63</summary>
        byte bitIndex;

        /// <summary>
        /// Reads bits and moves forward.
        /// </summary>
        /// <param name="bits">Maximum <c>64</c>.</param>
        /// <returns>Bits are placed on the right side of output.<br/>Ex. ...<c>0000000</c>(You are here)<c>1100101</c>... with <c>bits</c>=<c>5</c> returns <c>0b0000000000000000000000000000000000000000000000000000000000011001</c>.</returns>
        ulong ReadBits(byte bits = 8)
        {
            if (bits == 0) return 0;
            ulong output = 0;

            if (bitIndex + bits >= 64)
            {
                // read in 2 parts
                byte bits_1 = (byte)(64 - bitIndex);
                byte bits_2 = (byte)(bits - bits_1);

                ulong reading_1 = cache >> (64 - bits_1);

                byteIndex += 8;
                ObtainCache();

                if (bits_2 > 0)
                {
                    ulong reading_2 = cache >> (64 - bits_2);
                    cache <<= bits_2;
                    output = (reading_1 << bits_2) | reading_2;
                }
                else
                    output = reading_1;
            }
            else
            {
                // the read fits into cache
                output = cache >> (64 - bits);
                cache <<= bits;
            }

            bitIndex += bits;
            bitIndex %= 64;
            return output;
        }
        unsafe void ObtainCache()
        {
            if (byteIndex + 8 <= buf.Length) // main branch (here 99.9% of the time)
                cache = BinaryPrimitives.ReadUInt64BigEndian(buf.AsSpan((int)byteIndex, 8));

            // we are at the end of the byte[]; careful not to overextend the span
            else
            {
                cache = 0;
                uint idx = byteIndex;
                while (idx < buf.Length)
                {
                    cache <<= 8;
                    cache |= buf[idx];
                    idx++;
                }
                // move cache to the left
                cache <<= (byte)(64 - 8 * (buf.Length - byteIndex));
            }
        }
        [MethodImpl(AggressiveInlining)] public void Read(out sbyte value) => value = (sbyte)ReadBits(8);
        [MethodImpl(AggressiveInlining)] public void Read(out byte value) => value = (byte)ReadBits(8);

        [MethodImpl(AggressiveInlining)] public void Read(out short value) => value = (short)ReadBits(16);
        [MethodImpl(AggressiveInlining)] public void Read(out ushort value) => value = (ushort)ReadBits(16);

        [MethodImpl(AggressiveInlining)] public void Read(out int value) => value = (int)ReadBits(32);
        [MethodImpl(AggressiveInlining)] public void Read(out uint value) => value = (uint)ReadBits(32);

        [MethodImpl(AggressiveInlining)] public void Read(out long value) => value = (long)ReadBits(64);
        [MethodImpl(AggressiveInlining)] public void Read(out ulong value) => value = (ulong)ReadBits(64);

        [MethodImpl(AggressiveInlining)] public void Read(out bool value) => value = ReadBits(1) == 1;

        [MethodImpl(AggressiveInlining)] public void Read(out float value) => value = BitConverter.Int32BitsToSingle((int)ReadBits(32));
        [MethodImpl(AggressiveInlining)] public void Read(out double value) => value = BitConverter.Int64BitsToDouble((int)ReadBits(64));

        // [MethodImpl(AggressiveInlining)]
        // public void Read(out Vector2 value)
        // {
        //     Read(out float x);
        //     Read(out float y);
        //     value = new Vector2(x, y);
        // }
        // public void Read(out Fix64 value) => value = Fix64.FromRaw((long)ReadBits(64));


        [MethodImpl(AggressiveInlining)]
        public unsafe void Read<TE>(out TE value) where TE : unmanaged, Enum
        {
            switch (sizeof(TE))
            {
                case 1:
                {
                    Read(out byte v);
                    value = *(TE*)&v;
                    return;
                }
                case 2:
                {
                    Read(out ushort v);
                    value = *(TE*)&v;
                    return;
                }
                case 4:
                {
                    Read(out uint v);
                    value = *(TE*)&v;
                    return;
                }
                case 8:
                {
                    Read(out ulong v);
                    value = *(TE*)&v;
                    return;
                }
                default:
                {
                    throw new System.Exception();
                }
            }
        }

        ///<summary>Read any unmanaged struct (size never changes; no List, only fixed arrays).</summary>
        [MethodImpl(AggressiveInlining)]
        public unsafe void ReadStruct<TS>(out TS value) where TS : unmanaged
        {
            Unsafe.SkipInit(out value);
            byte* ptr = (byte*)System.Runtime.CompilerServices.Unsafe.AsPointer(ref value);
            ReadByteArrayUnsafe(ptr, sizeof(TS));
        }
        [MethodImpl(AggressiveInlining)]
        public unsafe void ReadString(out string s)
        {
            Read(out ushort len);
            if (len == ushort.MaxValue)
            {
                s = null;
                return;
            }
            if (len == 0)
            {
                s = string.Empty;
                return;
            }

            byte[] bytes = new byte[len];

            fixed (byte* ptr = bytes)
            {
                byte* p = ptr;
                ReadByteArrayUnsafe(p, bytes.Length);
            }

            s = System.Text.Encoding.UTF8.GetString(bytes);
        }

        /// <summary>
        /// Initially not public API, don't use this unless you know what you are doing!<br/>
        /// Requires a managed reference to be "pinned" to use pointers. If you have a <c>byte[]</c>, use a <c>fixed(byte* ptr = bytes)</c> statement.
        /// </summary>
        [MethodImpl(AggressiveInlining)]
        public unsafe void ReadByteArrayUnsafe(byte* ptr, int len)
        {
            int remaining = len;

            while (remaining >= 8)
            {
                BinaryPrimitives.WriteUInt64BigEndian(new Span<byte>(ptr, 8), ReadBits(64));
                ptr += 8;
                remaining -= 8;
            }
            while (remaining > 0)
            {
                *ptr = (byte)ReadBits(8);
                ptr += 1;
                remaining -= 1;
            }
        }
    }

    public class ByteList
    {
        public byte[] backingArr;
        public int Count;

        [MethodImpl(AggressiveInlining)] public ByteList(int capacity = 0) { backingArr = new byte[capacity]; Count = 0; }

        [MethodImpl(AggressiveInlining)]
        public void EnsureCapacity(int needed)
        {
            if (needed > backingArr.Length) Array.Resize(ref backingArr, Mathb.Max(backingArr.Length == 0 ? 4 : backingArr.Length * 2, needed));
        }

        [MethodImpl(AggressiveInlining)]
        public void Add(byte value)
        {
            EnsureCapacity(Count + 1);
            backingArr[Count++] = value;
        }
        [MethodImpl(AggressiveInlining)]
        public unsafe void Add_u64(ulong value)
        {
            EnsureCapacity(Count + 8);
            fixed (byte* p = &backingArr[Count])
                *(ulong*)p = BinaryPrimitives.ReverseEndianness(value);
            Count += 8;
        }
        [MethodImpl(AggressiveInlining)]
        public void AddRange(ReadOnlySpan<byte> values)
        {
            EnsureCapacity(Count + values.Length);
            values.CopyTo(backingArr.AsSpan(Count));
            Count += values.Length;
        }

        public byte this[int index]
        {
            [MethodImpl(AggressiveInlining)]
            get => backingArr[index];
            [MethodImpl(AggressiveInlining)]
            set => backingArr[index] = value;
        }
        [MethodImpl(AggressiveInlining)]
        public byte[] ToArray()
        {
            byte[] result = new byte[Count];
            Array.Copy(backingArr, result, Count);
            return result;
        }

        [MethodImpl(AggressiveInlining)] public Span<byte> AsSpan() => backingArr.AsSpan(0, Count);
    }
}